targets/csharp/source/source/Uunit/PluginManagerTest.cs

```diff
@@ -0,0 +1,125 @@
+{
+    public class PluginManagerTest : UUnitTestCase
+    {
+        private class CustomSerializerPlugin : ISerializerPlugin
+        {
+            public T DeserializeObject<T>(string serialized)
+            {
+                throw new NotImplementedException();
+            }
+
+            public T DeserializeObject<T>(string serialized, object serializerStrategy)
+            {
+                throw new NotImplementedException();
+            }
+
+            public object DeserializeObject(string serialized)
+            {
+                throw new NotImplementedException();
+            }
+
+            public string SerializeObject(object obj)
+            {
+                throw new NotImplementedException();
+            }
+
+            public string SerializeObject(object obj, object serializerStrategy)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private class CustomTransportPlugin : ITransportPlugin
+        {
+            public string Name;
+
+            public async Task<object> DoPost(string urlPath, object request, Dictionary<string, string> headers)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        /// <summary>
+        /// CLIENT AND SERVER API
+        /// Test that plugin manager returns default plugins if they are not set.
+        /// </summary>
+        [UUnitTest]
+        public void PluginManagerDefaultPlugins(UUnitTestContext testContext)
+        {
+            var playFabSerializer = PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer);
+            var playFabTransport = PluginManager.GetPlugin<ITransportPlugin>(PluginContract.PlayFab_Transport);
+
+            testContext.NotNull(playFabSerializer);
+            testContext.NotNull(playFabTransport);
+            testContext.EndTest(UUnitFinishState.PASSED, null);
+        }
+
+        /// <summary>
+        /// CLIENT AND SERVER API
+        /// Test that plugin manager can set and return a custom plugin.
+        /// </summary>
+        [UUnitTest]
+        public void PluginManagerCustomPlugin(UUnitTestContext testContext)
+        {
+            var playFabSerializer = PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer);
+            var customSerializer = new CustomSerializerPlugin();
+            try
+            {
+                // Set a custom serializer plugin
+                PluginManager.SetPlugin(customSerializer, PluginContract.PlayFab_Serializer);
+
+                // Get serializer plugin from manager
+                var serializerPlugin = PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer);
+
+                // Verify
+                testContext.True(object.ReferenceEquals(serializerPlugin, customSerializer));
+                testContext.EndTest(UUnitFinishState.PASSED, null);
+            }
+            catch (Exception e)
```

Comment:
What manner of exception is possible here?
Is an exception part of the normal customer experience?

Reply:
I like tests to match the customer's experience as much as possible, because we should be testing that the customer's experience matches what we expect.

Reply:
The finally block below is needed in case if anything goes wrong (including any exceptions that may happen inside PluginManager - in the future, if not now - we can't guarantee it will never happen). But you are right, the catch block itself seems redundant since the test will fail on exception anyway

---
In reply to: [202453615](https://github.com/PlayFab/SDKGenerator/pull/175#discussion_r202453615) [](ancestors = 202453615)